### PR TITLE
implement df command

### PR DESCRIPTION
**Pull request by framlog**
See #431

I implement a basic df command, which can work in macos and unix.

### CODE REVIEW

File: src/df/df.rs [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+    ffree: u64
+}
+
+#[derive(Debug, Clone)]
+struct Filesystem {
+    mountinfo: MountInfo,
+    usage: FsUsage,
+}
+
+macro_rules! exit_with_error(
```
Comment (Arcterus) [inline src/df/df.rs]: I believe this does the same thing as the `crash!` macro.

File: src/df/df.rs [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+        let mut mptr: *mut statfs = ptr::null_mut();
+        let len = unsafe { getmntinfo(&mut mptr, 1 as c_int) };
+        if len < 0 {
+            exit_with_error!("getmntinfo() failed");
+        }
+        let mounts = unsafe { slice::from_raw_parts(mptr, len as usize) };
+        return mounts.iter().map(|m| m.to_mount_info()).collect::<Vec<_>>();
+    }
+    // panic for other os
+    exit_with_error!("unimplemented");
```
Comment (Arcterus) [inline src/df/df.rs]: Should prob just be `unimplemented!()`

File: src/df/df.rs [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+static OPT_EXCLUDE_TYPE: &str = "exclude-type";
+static OPT_VERSION: &str = "version";
+
+static MOUNT_OPT_BIND: &str = "bind";
+
+#[cfg(any(target_os = "freebsd", target_os = "macos"))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct fsid_t {
```
[the remark this answers is not in the document]
Comment (framlog) [reply src/df/df.rs]: Nope, I forgot to grep it in libc merely.

File: src/df/df.rs (hunk 1 above)
Comment (framlog) [reply src/df/df.rs]: yep, I didn't know that...

**Comment by framlog** [comment]
> I'm a little worried with how much of this is OS-dependent (and architecture-dependent?). It's fine if there aren't any crates that handle some of this though.

@Arcterus I didn't find crates that work on this. But I think maybe we could abstract some code to do so since many commands have already did it on their own.

File: src/df/df.rs [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+    mount_root: String,
+    remote: bool,
+    dummy: bool,
+}
+
+#[cfg(all(target_os = "macos", target_arch = "x86_64"))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct statfs {
```
Comment (Arcterus) [inline src/df/df.rs]: I believe this is available in `libc` as well.

File: src/df/df.rs [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+    f_mntonname: [c_char; 1024usize],
+    f_mntfromname: [c_char; 1024usize],
+    f_reserved: [u32; 8usize],
+}
+
+#[cfg(all(target_os = "freebsd", not(all(target_os = "macos", target_arch = "x86_64"))))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct statfs {
```
Comment (Arcterus) [inline src/df/df.rs]: I'm not totally sure this version of the struct is available, but it might be.

**Comment by Arcterus** [comment]
It's fine if there aren't any crates.  Can you try extracting some of the code that might be useful for other utilities to `uucore`?

File: src/df/df.rs (hunk 4 above)
Comment (framlog) [reply src/df/df.rs]: libc does have a `statfs` struct, but it's missing some fields which I need to use here.

Comment (Arcterus) [reply src/df/df.rs]: Is it missing fields for Darwin?  It looked like everything was there, but perhaps I missed something.  It does seem to be missing needed fields for FreeBSD though.

Comment (framlog) [reply src/df/df.rs]: No, that struct is complete for Darwin. And there's no need to rewrite it. sry that i didn't check that before.

**Comment by sylvestre** [comment]
worked moved in https://github.com/uutils/coreutils/pull/1381

**Pull request closed without merging**